profiles/killbill/src/main/java/org/killbill/billing/server/modules/KillBillShiroWebModule.java

```diff
@@ -91,9 +92,9 @@ public String getString(final String propertyName) {
 
         // Magic provider to configure the cache manager
         if (redisCacheConfig.isRedisCachingEnabled()) {
-            bind(CacheManager.class).toProvider(RedisShiroManagerProvider.class).asEagerSingleton();
+            bind(CacheManager.class).annotatedWith(Names.named(SHIRO_CACHE_MANAGER)).toProvider(RedisShiroManagerProvider.class).asEagerSingleton();
         } else {
-            bind(CacheManager.class).toProvider(EhcacheShiroManagerProvider.class).asEagerSingleton();
+            bind(CacheManager.class).annotatedWith(Names.named(SHIRO_CACHE_MANAGER)).toProvider(EhcacheShiroManagerProvider.class).asEagerSingleton();
```

Comment:
This is a `org.apache.shiro.cache.CacheManager`, not a `javax.cache.CacheManager` -- I don't think I understand how it causes #1408.

Are you sure it's not related to the fact that `MultiTenantEventConfig` is a just-in-time binding. Did you try `bind(EventConfig.class).to(MultiTenantEventConfig.class).asEagerSingleton()`?

Reply:
That's precisely the problem: We have two bindings for the same interface `CacheManager` that bind to 2 different classes -> `ClassCastException`. Fails consistently before fix , and works fine afterwards.